src/types.jl

```diff
@@ -53,6 +53,12 @@ function breakpointchar(bps::BreakpointState)
     return bps.condition === falsecondition ? ' ' : 'd'     # no breakpoint : disabled
 end
 
+mutable struct DispatchableMethod
+    next::Union{Nothing,DispatchableMethod}  # linked-list representation
+    frameinstance::Any # really a FrameInstance but we have a cyclic dependency
+    sig::Type # for speed of matching, this is a *concrete* signature. `sig <: frameinstance.framecode.scope.sig`
```

Comment:
If you're doing this, can you implement the `compiled` idea too? With this diff
```diff
tim@diva:~/.julia/dev/JuliaInterpreter$ git diff
diff --git a/src/construct.jl b/src/construct.jl
index 9b70786..796c4ab 100644
--- a/src/construct.jl
+++ b/src/construct.jl
@@ -236,6 +236,7 @@ function prepare_call(@nospecialize(f), allargs; enter_generated = false)
     end
     argtypesv = Any[_Typeof(a) for a in allargs]
     argtypes = Tuple{argtypesv...}
+    println("Calling which (super-slow)")
     method = whichtt(argtypes)
     if method === nothing
         # Call it to generate the exact error
```
and the following test code
```julia
using JuliaInterpreter

bar(::Int) = 1
bar(::Float64) = 2
push!(JuliaInterpreter.compiled_methods, @which bar(1.0))

function foo(A)
    s = 0
    for a in A
        @bp
        s += bar(a)
    end
    return s
end
A = Real[1, 1, 1, 1, 1.0, 1.0, 1.0, 1.0]
frame = JuliaInterpreter.enter_call(foo, A)
for i = 1:length(A)
    println("iteration ", i)
    debug_command(frame, :c)
    for i = 1:3
        debug_command(frame, :se)  # step forward to the call to bar
    end
    # display(frame.framecode.methodtables)
end
```
I think you'll see what I'm worried about. (Note that the `Int` method of `bar` is recursively interpreted and the `Float64` one in `Compiled` mode.) If you inspect `frame.framecode.methodtables` you'll see that `Compiled` methods can't exploit the local method tables.

In the longer run we want to store the `MethodInstance`, and then use a `ccall` to skip the `jl_lookup_generic` from [`jl_apply_generic`](https://github.com/JuliaLang/julia/blob/5aea4569e988c6e231a2d9204c7e39193b961891/src/gf.c#L2206-L2214). That would basically mean DLLEXPORTing [this](https://github.com/JuliaLang/julia/blob/5aea4569e988c6e231a2d9204c7e39193b961891/src/gf.c#L2037-L2051). That will eliminate the runtime dispatch on compiled dispatch from the interpreter.

Reply:
Correction: there's already [this](https://github.com/JuliaLang/julia/blob/5aea4569e988c6e231a2d9204c7e39193b961891/src/gf.c#L2053-L2057), so we don't have to modify Julia at all.

Reply:
I think I am doing something wrong heh

```
julia> mi = Core.Compiler.specialize_method(@which(sin(2.0)), Tuple{typeof(sin), Float64}, Core.svec(Float64))
MethodInstance for sin(::Float64)

julia> args = [2.0]
1-element Array{Float64,1}:
 2.0

julia> ccall(:jl_invoke, Any, (Any, Vector{Any}, Cuint, Any), sin, args, 1, mi)
2.255484558e-314

julia> sin(args...)
0.9092974268256817
```

Reply:
Anyway, I added caching for Compiled calls as well.

Reply:
Ok, `args` should be a `Vector{Any}` and the args type should be `Ptr{Any}`.

Reply:
```jl
function jl_generic(args)
    +(args...)
end

function jl_invoke(args, mi)
     ccall(:jl_invoke, Any, (Any, Ptr{Any}, Cuint, Any), +, args, 2, mi)
end

args = Any[1, 2]
@btime jl_generic($args)
# 42 ns
mi = Core.Compiler.specialize_method(@which(1 + 1), Tuple{typeof(+), Int, Int}, Core.svec())
@btime foo($args, $mi)
# 11 ns
```

Reply:
Not quite sure I get what `foo` is, but if you're testing `jl_invoke` against run-time dispatch keep in mind that the timings are rather dependent on how many methods there are. `convert` might be in a different category :smile:.

Reply:
Sorry, had a typo, `foo` should have been `jl_invoke`. Just wanted to show that using `jl_invoke` with a MethodInstance was faster :)